File: packages/license/src/keyring.ts
import { b64uDecode } from "./token";

export type KeyStatus = "active" | "retired" | "compromised";

export type TrustedKey = {
  readonly kid: string;
  readonly alg: "Ed25519";
  readonly publicKey: Uint8Array;
  readonly validFrom: number;
  readonly validUntil: number | null;
  readonly status: KeyStatus;
};

function key(
  kid: string,
  hex: string,
  validFrom: number,
  validUntil: number | null,
  status: KeyStatus,
): TrustedKey {
  const bytes = new Uint8Array(32);
  for (let i = 0; i < 32; i++)
    bytes[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  return Object.freeze({
    kid,
    alg: "Ed25519",
    publicKey: bytes,
    validFrom,
    validUntil,
    status,
  });
}

export const TUNNET_TRUSTED_KEYS: readonly TrustedKey[] = Object.freeze([
  key(
    "tnk-2025-01",
    "54544bc6251b8076e7cdceff6b741d258b37a6a93020a03a251fe209b325ebbd",
    0,
    null,
    "active",
  ),
]);

export class Keyring {
  readonly #byKid: ReadonlyMap<string, TrustedKey>;

  constructor(keys: readonly TrustedKey[] = TUNNET_TRUSTED_KEYS) {
    if (keys.length === 0)
      throw new Error("keyring must contain at least one key");
    const map = new Map<string, TrustedKey>();
    for (const k of keys) {
      if (k.publicKey.length !== 32)
        throw new Error(`key ${k.kid}: public key must be 32 bytes`);
      if (map.has(k.kid)) throw new Error(`duplicate kid: ${k.kid}`);
      map.set(k.kid, k);
    }
    this.#byKid = map;
  }

  static fromSpkiBase64(kid: string, spki: string): TrustedKey {
    const der = b64uDecode(
      spki.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, ""),
    );
    if (der.length !== 44) throw new Error("unexpected Ed25519 SPKI length");
    return Object.freeze({
      kid,
      alg: "Ed25519" as const,
      publicKey: der.slice(12),
      validFrom: 0,
      validUntil: null,
      status: "active" as const,
    });
  }

  get(kid: string): TrustedKey | undefined {
    return this.#byKid.get(kid);
  }
}

File: crates/tunnet-license/src/keyring.rs
//! Trusted signing keys for tnlic1 tokens.

use std::collections::HashMap;

use jiff::Timestamp;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum KeyStatus {
    Active,
    Retired,
    Compromised,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TrustedKey {
    pub kid: String,
    pub public_key: [u8; 32],
    pub valid_from: Timestamp,
    pub valid_until: Option<Timestamp>,
    pub status: KeyStatus,
}

impl TrustedKey {
    pub const ALG: &'static str = "Ed25519";

    pub fn from_hex(
        kid: impl Into<String>,
        hex: &str,
        valid_from: Timestamp,
        valid_until: Option<Timestamp>,
        status: KeyStatus,
    ) -> Result<Self, String> {
        let bytes = hex::decode(hex).map_err(|e| e.to_string())?;
        if bytes.len() != 32 {
            return Err(format!("public key must be 32 bytes, got {}", bytes.len()));
        }
        let mut public_key = [0u8; 32];
        public_key.copy_from_slice(&bytes);
        Ok(Self {
            kid: kid.into(),
            public_key,
            valid_from,
            valid_until,
            status,
        })
    }
}

/// Built-in Tunnet production/dev keyring (`tnk-2025-01`).
pub fn tunnet_trusted_keys() -> Vec<TrustedKey> {
    vec![
        TrustedKey::from_hex(
            "tnk-2025-01",
            "54544bc6251b8076e7cdceff6b741d258b37a6a93020a03a251fe209b325ebbd",
            Timestamp::UNIX_EPOCH,
            None,
            KeyStatus::Active,
        )
        .expect("static key hex"),
    ]
}

/// Lazy static list matching TS `TUNNET_TRUSTED_KEYS`.
pub static TUNNET_TRUSTED_KEYS: std::sync::LazyLock<Vec<TrustedKey>> =
    std::sync::LazyLock::new(tunnet_trusted_keys);

#[derive(Debug, Clone)]
pub struct Keyring {
    by_kid: HashMap<String, TrustedKey>,
}

impl Keyring {
    pub fn new(keys: impl IntoIterator<Item = TrustedKey>) -> Result<Self, String> {
        let mut by_kid = HashMap::new();
        for k in keys {
            if by_kid.contains_key(&k.kid) {
                return Err(format!("duplicate kid: {}", k.kid));
            }
            by_kid.insert(k.kid.clone(), k);
        }
        if by_kid.is_empty() {
            return Err("keyring must contain at least one key".into());
        }
        Ok(Self { by_kid })
    }

    pub fn default_tunnet() -> Self {
        Self::new(TUNNET_TRUSTED_KEYS.iter().cloned()).expect("default keyring")
    }

    pub fn get(&self, kid: &str) -> Option<&TrustedKey> {
        self.by_kid.get(kid)
    }
}

impl Default for Keyring {
    fn default() -> Self {
        Self::default_tunnet()
    }
}
